workflows/sweep_simulate.snake

```diff
@@ -0,0 +1,324 @@
+demo_model = config["demo_model"]
+
+# Simulate across multiple DFEs/annotations
+dfe_list = config["dfe_list"]
+annotation_list = config["annotation_list"]
+
+# Region size 
+region_size = config["region_size"]
+
+def genetic_distance(ratemap, x, x0=0):
+    """
+    Function that returns the genetic distance (in cM) between positions x and 
+    x0 based on a msprime.RateMap with recombination rates.
+    """
+    return 100*(ratemap.get_cumulative_mass(x) - ratemap.get_cumulative_mass(x0))
+
+
+def make_buffer(gmap, cM, left, right):
+    """
+    This function adds a buffer zone to the interval [left, right] of fixed genetic distance cM
+    and returns the physical positions of the buffered interval.
+    The buffered intervals are bounded by the size of the genetic map gmap.
+    """
+    assert right > left
+    assert left >= 0
+    assert right <= gmap.sequence_length
+    # this is gonna fail in the beginning and ends of chroms bc of nans and 0s in ratemap
+    try:
+        left = root_scalar(
+            lambda y: genetic_distance(ratemap=gmap, x=left, x0=y) - cM,
+            method='bisect',
+            bracket=[0, gmap.sequence_length],
+        )
+        left = left.root
+    except:
+        left = 0
+    try:
+        right = root_scalar(
+            lambda y: genetic_distance(ratemap=gmap, x=y, x0=right) - cM,
+            method='bisect',
+            bracket=[0, gmap.sequence_length],
+        )
+        right = right.root
+    except:
+        right = gmap.sequence_length
+    return left, right
+
+def add_window_metadata(ts, window_left, window_right, window_bleft, window_bright):
+    tables = ts.dump_tables()
+    metadata = tables.metadata
+    new_metadata = {
+        "windowing":
+            {
+                "window_left": window_left,
+                "window_bleft": window_bleft,
+                "window_right": window_right,
+                "window_bright": window_bright,
+            }
+    }
+    metadata.update(new_metadata)
+    tables.metadata = metadata
+    return tables.tree_sequence()
+
+def compute_stats(ts):
+    offset = ts.metadata['windowing']['window_bleft']
+    windows_bpoints = np.array([ts.metadata['windowing']['window_bleft']-offset,
+                                ts.metadata['windowing']['window_left']-offset,
+                                ts.metadata['windowing']['window_right']-offset,
+                                ts.metadata['windowing']['window_bright']-offset], dtype=np.int32)
+    # removing duplicates if bleft and left are the same or brighta and right
+    windows_bpoints = np.unique(windows_bpoints)
+    # something wonky happening with float conversion
+    if windows_bpoints[-1] == ts.sequence_length - 1:
+        windows_bpoints[-1] = ts.sequence_length
+    assert (windows_bpoints.shape == (2,)) or windows_bpoints.shape == (4,)
+    taj=ts.Tajimas_D(windows=windows_bpoints,mode="site")
+    focal_taj = taj[len(taj)//2]
+    return focal_taj, taj
```

Comment:
i think these function should be put in a module, perhaps named `sweep_utils.py`. it's useful to separate the code from the workflow.